[FIX] Rewrite inheritance principle (#946)
bids-standard/bids-specification: src/02-common-principles.md

## The Inheritance Principle

> Any metadata file (such as `.json`, `.bvec` or `.tsv`) MAY be defined at any
> directory level. For any given data file, any metadata file at that directory
> level or higher that does not include any entities absent from the name of the
> data file and possesses the same suffix are applicable to that data file. Such
> files are loaded from the top of the directory hierarchy downwards, such that
> values from the top level are inherited by all data files at lower levels to
> which it is applicable unless overridden by a value for the same key present
> in another metadata file at a lower level (though it is RECOMMENDED to minimise
> the extent of such overrides). There is no notion of "unsetting" a
> key/value pair.

Reviewer: ```suggestion
Any metadata file (such as `.json`, `.bvec` or `.tsv`) MAY be defined at any directory level. 
For a given data file, any metadata file at that directory level or higher 
is applicable to that data file if:
- the metadata file and the data file possess the same suffix,
- the metadata file does not include any entities absent from the name of the data file. 
Such files are loaded from the top of the directory hierarchy downwards, 
such that values from the top level are inherited by all data files 
at lower levels to which it is applicable unless overridden 
by a value for the same key present in another metadata file at a lower level 
(though it is RECOMMENDED to minimise the extent of such overrides). 
There is no notion of "unsetting" a key/value pair.
```

Author: Yes, I'd contemplated bullet-pointing this myself; or potentially even enumerating so that datasets violating specific requirements / software code can be cross-referenced more precisely?

@VisLab: Good catch. There's going to need to be a hard split between JSON and other metadata files. For `.bvec` / `.bval` / `.tsv`, it I think needs to be simply selection of whichever applicable file is lowest in the directory tree; anything above it gets ignored, there's no reason to describe the contents of such as being "overwritten" by the content of files lower down. Only edge case that comes to mind would be `.tsv` files that happen to contain exactly the same columns (including their titles), but even then it's hard to justify choosing between a concatenation and a replacement. Are there any other metadata types that are worth being familiar with for the sake of getting the language / logic of the inheritance principle right?

Reviewer: Good idea about using an ordered list and referring to those points in the example.

Reviewer 2: So am I understanding that for JSON the resolution is top-down, with top (closest to the root) having precedence.
For the others it is bottom up, with the bottom (farthest from the root) having precedence?  

The notions of "lowest" and "highest" level in the tree isn't necessarily clear.  Does lowest mean farthest from the root? 

I honestly thought only JSON files got inherited in the sense that there could be multiple applicable ones for the same suffix.

Reviewer: I think the only case of "official" of TSV inheritance we have around is having a single `task-foo_events.tsv` in the root of the directory in case all subjects / runs had the same design:

see from the BIDS examples

```bash
╰─⠠⠵ ls -l */*task*tsv
-rw-rw-r-- 1 remi remi  143 Nov 10 11:33 ds114/task-covertverbgeneration_events.tsv
-rw-rw-r-- 1 remi remi  280 Nov 10 11:33 ds114/task-fingerfootlips_events.tsv
-rw-rw-r-- 1 remi remi  143 Nov 10 11:33 ds114/task-overtverbgeneration_events.tsv
-rw-rw-r-- 1 remi remi  127 Nov 10 11:33 ds114/task-overtwordrepetition_events.tsv
-rw-rw-r-- 1 remi remi 1054 Nov 10 11:33 eeg_ds000117/task-facerecognition_channels.tsv
-rw-rw-r-- 1 remi remi 1807 Nov 10 08:42 synthetic/task-nback_events.tsv
```

Reviewer 2: We probably need to expand the discussion of events inheritance.  Inherited `events.tsv files` are needed at several levels.  

For example, because BIDS forces multi-modality data to be separated into separate directories, simultaneously recorded MEG/EEG or fMRI/EEG will have the same event files for each run.  

In the following example from [hed-examples](https://github.com/hed-standard/hed-examples/tree/main/datasets/eeg_ds003654s_inheritance), EEG and MEG were recorded simultaneously, but separated into separate modality directories.  

```
ds_003654s:
       task-FacePerception_events.json
       sub-02:
               sub-002_task-FacePerception_events.json
               sub-002_task-FacePerception_run-1_events.tsv
               eeg:
                      sub-002_task-FacePerception_run-1_eeg.set 
               meg:
                      sub-002_task-FacePerception_run-1_meg.fif
   . . .
```

Right now we are assuming that multiple `events.json` files can be applicable to a given `events.tsv` files and that the keys are resolved from directory root down in the JSON files.  I think this is the right behavior for JSON files, but it needs to be clarified in the specification.

We have only made a single `events.tsv`  applicable to any given recording, but that `events.tsv` file could appear anywhere in the directory hierarchy.   

Events are a special case, and I think it might be feasible/useful to allow multiple events files at different levels to apply to a given run. The events are the union of the rows/join of columns.  This might be too complicated.  

The issue also comes up with derivatives.  Here the derivatives might have their own set of events. If the events from the raw data also apply do they need to always be copied into the events files? There was some discussion of this in the past, but there didn't seem to be a clear resolution or consensus on this when the discussion occurred.

Reviewer 2: I also have a general question about inheritance.  Is the matching of entity components and their names case sensitive?

Author: > The notions of "lowest" and "highest" level in the tree isn't necessarily clear. Does lowest mean farthest from the root?

I always have to cross-reference elsewhere in the spec to remind myself which is which. Does anyone know if there's a robust reference-able definition somewhere? Or does this need to be defined upfront in the spec?

Reviewer: > Is the matching of entity components and their names case sensitive?

I think the case sensitivity issues was solved in https://github.com/bids-standard/bids-specification/pull/858

}
```

## The Inheritance Principle

1.  Any metadata file (such as `.json`, `.bvec` or `.tsv`) MAY be defined at any directory level.

> 1.  For a given data file, any metadata file is applicable to that data file if:

Reviewer 3: although markdown does numbering for us, since there are direct references to them in the text (e.g. `(rule 5.2.2)`) I think it would be better just to number them here as well, so it is easier to refer back while reading markdown and not rendered version

Author: Better would be enumeration cross-references; but as far as I can tell no standardized solution for such exists?

Author: > it would be better just to number them here as well

Not permitted: Lestropie/bids-specification#2

Reviewer 3: eh, thanks for trying it out! 
Indeed from https://www.npmjs.com/package/remark-lint-ordered-list-marker-value I see that
- the default is `"one"`
- there seems to be no I do not see some way to override locally in a file/for this specific list.
   - https://stackoverflow.com/a/63753061/1265472 suggests that a local to a directory `.remarkrc` could be used to overload some settings, but that would not help here/more confusing than helpful in the long run
- the other places in BIDS specification follow the "one" convention and there it seems to make sense

so I guess I will "retract" my invaluable comment ;-) Thanks again for trying

Reviewer 3: actually according to https://github.com/remarkjs/remark-lint#ignore-warnings there is an inline way to disable/re-enable specific warnings in a specific block, but I do not see there a way to change configuration setting (i.e. make it "ordered") . I will suggest this disable/reenable in the PR just to see if works -- not sure if worth it if we can't lint the order

    1.  It is stored at the same directory level or higher;
    1.  The metadata and the data filenames possess the same suffix;
    1.  The metadata filename does not include any entity absent from the data filename.

> 1.  A metadata file MUST NOT have a file name that would be otherwise applicable

Reviewer 3: ```suggestion
1.  A metadata file MUST NOT have a filename that would be otherwise applicable
```
if there would be agreement on https://github.com/bids-standard/bids-specification/pull/962 prompted by this line ;-)

    to some data file based on rules 2.2 and 2.3 but is made inapplicable based on its
    location in the directory structure as per rule 2.1.

1.  There MUST NOT be multiple metadata files applicable to a data file at one level

>     of the directory hierarchy.

Reviewer 3: FWIW: it is such a nice concise rule which helps to avoid ambiguity and the workaround in Example 3 is quite cute ;)

Reviewer 3: OH -- I remembered what bothered me in my original #102 and outlined in Edit 1 there. Citing an example from there:

> I placed myself into a corner with an example of having e.g.
> 
> ```
> sub-1_task-task1_run-1_bold.json and
> sub-1_task-task1_acq-X_run-1_bold.json
> ```
>
> per subject (should be ok), and then trying to aggregate over them while retaining also _acq- if defined. 

and there are many other entities/use-cases which would fall into similar situation.  What I am thinking is to add a clarification as subitem here or a separate rule:

- A metadata file at the data files level of hierarchy MUST not be considered for inheritance if there is a matching in entities data file.

This would make `sub-1_task-task1_run-1_bold.json` not a to be considered for providing metadata to enrich `sub-1_task-task1_acq-X_run-1_bold.json`, and examples below would stay valid (we might want to add an example for this rule though)

Author: > it is such a nice concise rule which helps to avoid ambiguity and the workaround in Example 3 is quite cute ;)

Well, actually (from first post):

> Full disclosure: as per #259, I would like to pursue the prospect of modifying the inheritance principle itself, not just the description of such; specifically removal of the preclusion of having multiple applicable JSONs at one level of the hierarchy.

😬 

The proposed changes here do at least I think demonstrate what would need to change in order to facilitate such. Any text I propose for such will need to be very clear for both users and software creators exactly what is permitted vs. not permitted and in what order JSONs should be loaded. But as I said in the original post, this is not a change in descriptive language but a change in prescriptive permissible structure, and so IMO necessitates either a minor or major version change rather than just a patch.

Reviewer 3: Gotcha. Given the currently loose formulation of the principle, depending on the changes to it, I think it might be feasible to get it done within minor revision. So it would be great to see this PR be finalized/merged soon.

}
```

## The Inheritance Principle

1.  Any metadata file (such as `.json`, `.bvec` or `.tsv`) MAY be defined at any directory level.

1.  For a given data file, any metadata file is applicable to that data file if:
    1.  It is stored at the same directory level or higher;
    1.  The metadata and the data filenames possess the same suffix;
    1.  The metadata filename does not include any entity absent from the data filename.

1.  A metadata file MUST NOT have a filename that would be otherwise applicable
    to some data file based on rules 2.2 and 2.3 but is made inapplicable based on its

>     location in the directory structure as per rule 2.1.

Reviewer 4: I like these back-references to 2.2, 2.3 etc. -> however in the rendered version, the nested enumerated list has letters, not numbers (see screenshot below)

![image](https://user-images.githubusercontent.com/9084751/146912892-e6911f75-95ff-4384-9891-93260f33fd2d.png)

So I would suggest to refer to 2a, 2b, 2c here.

Author: Closed by 8e808a12.

1.  There MUST NOT be multiple metadata files applicable to a data file at one level
    of the directory hierarchy.

1.  If multiple metadata files are applicable to a given data file:

    1.  For [tabular files](#tabular-files) and other simple metadata files

>         (for instance, [`bvec` / `bval` files for diffusion MRI](#04-modality-specific-files/01-magnetic-resonance-imaging#required-gradient-orientation-information),

Reviewer 4: ```suggestion
        (for instance, [`bvec` / `bval` files for diffusion MRI](#04-modality-specific-files/01-magnetic-resonance-imaging#required-gradient-orientation-information)),
```

Author: Closed by 8e808a12.

>         only the file lowest in the filesystem hierarchy SHALL be treated as being

Reviewer 4: could we use MUST instead of SHALL here? Simply because so far there is zero mention of SHALL in the spec, but lots of MUST - and I thought these were equivalent?

Author: I had thought intuitively that there was a distinction between MUST and SHALL, but that appears to be incorrect based on RFC2119. My thinking here was that it's entirely plausible that an app may have no interest in the content of a sidecar file, and just load an e.g. image in isolation, and therefore stating that such sidecar information MUST be loaded is to me unnecessarily dictatorial. It's more like: if you are loading a data file, *and* you are interested in the sidecar data associated with that file, *then* only the lowest sidecar file MUST be utilised.

Author: See if you're happy with 8e808a12.

Reviewer 4: yes, thank you - resolving.

        applicable file that is lowest in the filesystem hierarchy.

    1.  For [JSON files](#key-value-files-dictionaries):

        1.  Files are loaded from the top of the directory hierarchy downwards,
            such that values from the top level are inherited by all data files
            at lower levels to which it is applicable unless overridden
            by a value for the same key present in another metadata file at a lower level

>             (though it is RECOMMENDED to minimise the extent of such overrides).

Reviewer 4: nitpick: Are we using American or British spelling throughout the spec? I had assumed we use American spelling, and indeed there are two occurrences of `minimize` and none of  `minimise` in the spec currently. Perhaps we should add a CI step (in a different issue/PR) that checks for American/British spelling. Just to be consistent.

Reviewer: Not a bad idea, but I am not quite ready to write that PR. :-)

        1.  There is no notion of "unsetting" a key/value pair.

Corollaries:

1.  As per rule 3, metadata files applicable only to a specific participant / session
    MUST be defined in or below the directory corresponding to that participant / session;
    similarly, a metadata file that is applicable to multiple participants / sessions
    MUST NOT be placed within a directory corresponding to only one such participant / session.

1.  It is permissible for a single metadata file to be applicable to multiple data
    files at that level of the hierarchy or below. Where such metadata content is consistent
    across multiple data files, it is RECOMMENDED to store metadata in this
    way, rather than duplicating that metadata content across multiple metadata files.

Example 1: Demonstration of inheritance principle

{{ MACROS___make_filetree_example(
    {
    "sub-01": {
        "func": {
            "sub-01_task-rest_acq-default_bold.nii.gz": "",
            "sub-01_task-rest_acq-longtr_bold.nii.gz": "",
            "sub-01_task-rest_acq-longtr_bold.json": "",
            }
        },
    "task-rest_bold.json": "",
    }
) }}

> Contents of file "task-rest_bold.json":

Reviewer 4: ```suggestion
Contents of file `task-rest_bold.json`:
```

Author: Closed by 8e808a12.

```JSON
{
    "EchoTime": 40.0,
    "RepetitionTime": 1000.0
}
```

> Contents of file "sub-01/func/sub-01_task-rest_acq-longtr_bold.json":

Reviewer 4: ```suggestion
Contents of file `sub-01/func/sub-01_task-rest_acq-longtr_bold.json`:
```

Author: Closed by 8e808a12.

```JSON
{
    "RepetitionTime": 3000.0
}
```

When reading image `sub-01/func/sub-01_task-rest_acq-default_bold.nii.gz`, only
metadata file `task-rest_bold.json` is read; file
`sub-01/func/sub-01_task-rest_acq-longtr_bold.json` is inapplicable as it contains

> entity "`acq-longtr`" that is absent from the image path (rule 2.3). When reading image

Reviewer 4: Rule 2.c, see my comment above.

I see that below this would be "5.b.ii" instead of 5.2.2, so maybe we need to discuss my suggestion - as 5.b.ii is kind of ugly.

Author: Revised by 8e808a12, but not yet resolving in case the enumeration formatting needs to be discussed further.

Reviewer 4: Looking at it again, I don't find it ugly anymore :-) Thanks for making the change.

`sub-01/func/sub-01_task-rest_acq-longtr_bold.nii.gz`, metadata file
`task-rest_bold.json` at the top level is read first, followed by file
`sub-01/func/sub-01_task-rest_acq-longtr_bold.json` at the bottom level (rule 5.2.1);

> the value for field "`RepetitionTime`" is therefore overridden to the value 3000.0.

Reviewer 4: ```suggestion
the value for field "`RepetitionTime`" is therefore overridden to the value `3000.0`.
```

Author: Closed by 8e808a12.

        accessing metadata associated with a data file MUST consider only the
        applicable file that is lowest in the filesystem hierarchy.

    1.  For [JSON files](#key-value-files-dictionaries), key-values are loaded
        from files from the top of the directory hierarchy downwards, such that
        key-values from the top level are inherited by all data files at lower
        levels to which it is applicable unless overridden by a value for the
        same key present in another metadata file at a lower level
        (though it is RECOMMENDED to minimise the extent of such overrides).

Corollaries:

1.  As per rule 3, metadata files applicable only to a specific participant / session
    MUST be defined in or below the directory corresponding to that participant / session;
    similarly, a metadata file that is applicable to multiple participants / sessions
    MUST NOT be placed within a directory corresponding to only one such participant / session.

1.  It is permissible for a single metadata file to be applicable to multiple data
    files at that level of the hierarchy or below. Where such metadata content is consistent

>     across multiple data files, it is RECOMMENDED to store metadata in this

Reviewer 2: I'm okay with taking out the recommended here.

>     across multiple data files, it is RECOMMENDED to store metadata in this

Author: >  This is a best practice written in the language of a validatable rule.

1.  This seems to extend beyond the RFC2119 definition. Is this an established BIDS-specific interpretation that is documented somewhere?

2.  Would the same criticism not also apply to 5.2?

    > For [JSON files](#key-value-files-dictionaries), key-values are loaded from files from the top of the directory hierarchy downwards, such that key-values from the top level are inherited by all data files at lower levels to which it is applicable unless overridden by a value for the same key present in another metadata file at a lower level (though it is RECOMMENDED to minimise the extent of such overrides).

    Issuing a validator warning regarding the presence of individual key-value overrides may be unexpected.

>     across multiple data files, it is RECOMMENDED to store metadata in this

Reviewer 4: >  Is this an established BIDS-specific interpretation that is documented somewhere?

AFAIK this is not documented anywhere, but among maintainers we have often discussed that RECOMMENDED corresponds to a "warning" level in the validator, and MUST corresponds to an "error" level in the validator. The main point of this consideration is to not design a specification that we cannot reasonably validate.

Having that said, there are many cases in the spec where we RECOMMEND or require (MUST), where validation is currently not happening and might be difficult to implement --- as you correctly point out in your second point.

Personally I am fine with having some recommendations that we cannot "warn" about (see especially 5.b here, but maybe also the point raised by Chris.).

    way, rather than duplicating that metadata content across multiple metadata files.

1.  Where multiple applicable JSON files are loaded as per rule 5.2, key-values can
    only be overwritten by files lower in the filesystem hierarchy; the absence of
    a key-value in a later file does not imply the "unsetting" of that field
    (indeed removal of existing fields is not possible).

Example 1: Demonstration of inheritance principle

{{ MACROS___make_filetree_example(
    {
    "sub-01": {
        "func": {
            "sub-01_task-rest_acq-default_bold.nii.gz": "",
            "sub-01_task-rest_acq-longtr_bold.nii.gz": "",
            "sub-01_task-rest_acq-longtr_bold.json": "",
            }
        },
    "task-rest_bold.json": "",
    }
) }}

Contents of file `task-rest_bold.json`:

```JSON
{
    "EchoTime": 40.0,
    "RepetitionTime": 1000.0
}
```

Contents of file `sub-01/func/sub-01_task-rest_acq-longtr_bold.json`:

```JSON
{

>     "RepetitionTime": 3000.0

Author: Closed by 33cc25a.

}
```

When reading image `sub-01/func/sub-01_task-rest_acq-default_bold.nii.gz`, only
metadata file `task-rest_bold.json` is read; file
`sub-01/func/sub-01_task-rest_acq-longtr_bold.json` is inapplicable as it contains
entity "`acq-longtr`" that is absent from the image path (rule 2.c). When reading image
`sub-01/func/sub-01_task-rest_acq-longtr_bold.nii.gz`, metadata file
`task-rest_bold.json` at the top level is read first, followed by file
`sub-01/func/sub-01_task-rest_acq-longtr_bold.json` at the bottom level (rule 5.b);

> the value for field "`RepetitionTime`" is therefore overridden to the value `3000.0`.

Author: Closed by 33cc25a.

1.  A metadata file MUST NOT have a filename that would be otherwise applicable
    to some data file based on rules 2.b and 2.c but is made inapplicable based on its
    location in the directory structure as per rule 2.a.

1.  There MUST NOT be multiple metadata files applicable to a data file at one level
    of the directory hierarchy.

1.  If multiple metadata files satisfy criteria 2.a-c above:

    1.  For [tabular files](#tabular-files) and other simple metadata files
        (for instance, [`bvec` / `bval` files for diffusion MRI](#04-modality-specific-files/01-magnetic-resonance-imaging#required-gradient-orientation-information)),
        accessing metadata associated with a data file MUST consider only the
        applicable file that is lowest in the filesystem hierarchy.

    1.  For [JSON files](#key-value-files-dictionaries), key-values are loaded
        from files from the top of the directory hierarchy downwards, such that
        key-values from the top level are inherited by all data files at lower
        levels to which it is applicable unless overridden by a value for the
        same key present in another metadata file at a lower level

>         (though it is RECOMMENDED to minimise the extent of such overrides).

Reviewer 4: ```suggestion
        (though it is RECOMMENDED to minimize the extent of such overrides).
```

Corollaries:

1.  As per rule 3, metadata files applicable only to a specific participant / session
    MUST be defined in or below the directory corresponding to that participant / session;
    similarly, a metadata file that is applicable to multiple participants / sessions
    MUST NOT be placed within a directory corresponding to only one such participant / session.

1.  It is permissible for a single metadata file to be applicable to multiple data
    files at that level of the hierarchy or below. Where such metadata content is consistent
    across multiple data files, it is RECOMMENDED to store metadata in this
    way, rather than duplicating that metadata content across multiple metadata files.

> 1.  Where multiple applicable JSON files are loaded as per rule 5.2, key-values can

Reviewer 4: ```suggestion
1.  Where multiple applicable JSON files are loaded as per rule 5.b, key-values can
```
